accounts/graph.py

```diff
@@ -0,0 +1,79 @@
+import json
+import requests
+import msal
+
+from django.contrib.auth import get_user_model
+
+#!! Change this to load parameters from env
+config = json.load(open("parameters.json"))
+
+app = msal.ConfidentialClientApplication(
+    config["client_id"], authority=config["authority"],
+    client_credential=config["secret"],
+)
+
+def acquire_graph_access_token():
+    result = None
+
+    # Firstly, looks up a token from cache
+    # Since we are looking for token for the current app, NOT for an end user,
+    # notice we give account parameter as None.
+    result = app.acquire_token_silent(config["scope"], account=None)
+
+    if not result:
+        print("No suitable token exists in cache. Let's get a new one from AAD.")
+        result = app.acquire_token_for_client(scopes=config["scope"])
+
+    if "access_token" in result:
+        return result['access_token']
+    else:
+        print(result.get("error"))
+        print(result.get("error_description"))
+        print(result.get("correlation_id"))  # You may need this when reporting a bug
+        return None
+
+def search_users(q):
+    token = acquire_graph_access_token()
+    results = requests.get(
+        #config["endpoint"],
+        "https://graph.microsoft.com/v1.0/users",
+        headers={
+            'Authorization': f'Bearer {token}',
+            'ConsistencyLevel': 'eventual'
+        },
+        params={
+            '$select': 'givenName,surname,mail,mailNickname,employeeId,OnPremisesExtensionAttributes',
+            #'$search': f'displayName:{q}'
+            '$search': "\"displayName:" + q + "\""
```

Comment:
just out of curiosity, why is this not using a fstring? just for improved readability.

Reply:
Yeah, you can see the commented out line where I tried to. For some reason Graph was claiming the request was malformed when using that. It needed the extra escaped quotation marks, and I couldn't work out how to do that with the fstring, although there may very well be a way of doing that.